src/java/org/apache/cassandra/index/sai/memory/MemtableIndex.java

```diff
@@ -61,32 +93,157 @@ public boolean isEmpty()
+                                                         allocatedBytes -> {
+                                                             memtable.markExtraOnHeapUsed(allocatedBytes, opGroup);
+                                                             estimatedOnHeapMemoryUsed.add(allocatedBytes);
+                                                         },
+                                                         allocatedBytes -> {
+                                                             memtable.markExtraOffHeapUsed(allocatedBytes, opGroup);
+                                                             estimatedOffHeapMemoryUsed.add(allocatedBytes);
+                                                         });
         writeCount.increment();
-        estimatedMemoryUsed.add(ram);
-        return ram;
     }
 
     public RangeIterator search(Expression expression, AbstractBounds<PartitionPosition> keyRange)
     {
-        return index.search(expression, keyRange);
+        RangeConcatIterator.Builder builder = RangeConcatIterator.builder();
+
+        for (MemoryIndex index : rangeIndexes)
+        {
+            builder.add(index.search(expression, keyRange));
+        }
+        return builder.build();
     }
 
-    public Iterator<Pair<ByteComparable, PrimaryKeys>> iterator()
+    /**
+     * NOTE: returned data may contain partition key not within the provided min and max which are only used to find
+     * corresponding subranges. We don't do filtering here to avoid unnecessary token comparison. In case of JBOD,
+     * min/max should align exactly at token boundaries. In case of tiered-storage, keys within min/max may not
+     * belong to the given sstable.
+     *
+     * @param min minimum partition key used to find min subrange
+     * @param max maximum partition key used to find max subrange
+     *
+     * @return iterator of indexed term to primary keys mapping in sorted by indexed term and primary key.
+     */
+    public Iterator<Pair<ByteComparable, Iterator<PrimaryKey>>> iterator(DecoratedKey min, DecoratedKey max)
     {
-        return index.iterator();
+        int minSubrange = min == null ? 0 : boundaries.getShardForKey(min);
+        int maxSubrange = max == null ? rangeIndexes.length - 1 : boundaries.getShardForKey(max);
+
+        List<Iterator<Pair<ByteComparable, PrimaryKeys>>> rangeLists = new ArrayList<>(maxSubrange - minSubrange + 1);
+        for (int i = minSubrange; i <= maxSubrange; i++)
+            rangeLists.add(rangeIndexes[i].iterator());
+
+        return MergeIterator.get(rangeLists, (o1, o2) -> ByteComparable.compare(o1.left, o2.left, ByteComparable.Version.OSS41),
+                                 new PrimaryKeysMergeReducer(rangeIndexes.length, clusteringComparator));
+    }
+
+    private static class PrimaryKeysMergeReducer extends MergeIterator.Reducer<Pair<ByteComparable, PrimaryKeys>, Pair<ByteComparable, Iterator<PrimaryKey>>>
```

Comment:
Nipick: I noticed that `MergeIterator.Reducer` has a comment  referring to names that don't match the type names in the code exactly. Fortunately it is simple enough I understand what the author meant.

```java
    /** Accumulator that collects values of type A, and outputs a value of type B. */
    public static abstract class Reducer<In,Out>
```

Reply:
There are many instances of this I have found while looking at this ticket but feel that it is better to leave things like this while we are merging forwards.

Reply:
I would prefer not to change comments in code I have not touched only because of the impact of merging changes. I would prefer to leave those changes to the CEP where we can push the changes directly to OS. 

I have added comments to this class and have tidied it up a bit to make it more understandable.